framework/decode/screenshot_handler.cpp

```diff
@@ -38,10 +38,121 @@ static constexpr size_t kUnormIndex = 0;
+    // Vulkan image formats for util::ScreenshotFormat::kPng
+    { VK_FORMAT_R8G8B8A8_UNORM, VK_FORMAT_R8G8B8A8_SRGB }
 };
 
+inline void OutputImage(const std::string&     filename,
+                        util::ScreenshotFormat file_format,
+                        uint32_t               width,
+                        uint32_t               height,
+                        uint64_t               size,
+                        void*                  data)
+{
+    switch (file_format)
+    {
+        default:
+            GFXRECON_LOG_ERROR("Screenshot format invalid!  Expected BMP or PNG, falling back to BMP.");
+            // Intentional fall-through
+        case util::ScreenshotFormat::kBmp:
+            if (!util::imagewriter::WriteBmpImage(filename + ".bmp", width, height, size, data))
+            {
+                GFXRECON_LOG_ERROR("Screenshot could not be created: failed to write BMP file %s", filename.c_str());
+            }
+            break;
+#ifdef ENABLE_PNG_SCREENSHOT
+        case util::ScreenshotFormat::kPng:
+            // For PNG format, we have to force the alpha channel to opaque, otherwise we get
+            // image artifacts on most image viewers from the transparency for anything that was
+            // rendered last to the image using transparency (think tree leaves).
+            uint32_t* working_data = reinterpret_cast<uint32_t*>(data);
+            for (uint32_t y = 0; y < height; ++y)
+            {
+                for (uint32_t x = 0; x < width; ++x)
+                {
+                    working_data[y * width + x] |= 0xFF000000;
+                }
+            }
+            if (!util::imagewriter::WritePngImage(filename + ".png", width, height, size, data))
+            {
+                GFXRECON_LOG_ERROR("Screenshot could not be created: failed to write PNG file %s", filename.c_str());
+            }
+            break;
+#endif // ENABLE_PNG_SCREENSHOT
+    }
+}
+
+#if defined(MAX_SCREENSHOT_THREADS) && MAX_SCREENSHOT_THREADS > 1
+struct ScreenshotImageThreadData
+{
+    bool                   valid = false;
+    std::string            filename;
+    util::ScreenshotFormat file_format;
+    std::vector<uint8_t>   data;
+    uint32_t               width  = 0;
+    uint32_t               height = 0;
+    size_t                 size   = 0;
+};
+std::atomic_bool                              g_exiting;
+static std::vector<ScreenshotImageThreadData> g_screenshot_thread_info;
+std::mutex                                    g_screenshot_write_mutex;
+
+static void WriteImageThread(int32_t index)
+{
+    while (!g_exiting)
+    {
+        bool valid = false;
+        {
+            std::unique_lock<std::mutex> lock(g_screenshot_write_mutex);
+            valid = g_screenshot_thread_info[index].valid;
+        }
+        // Do it this way so we're not sleeping while holding onto the lock
+        if (!valid)
+        {
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
```

Comment:
I see you chose the consumers to periodically sleep when there's no work and the producer to busy wait if all consumers are busy. Do you think conditional variables would be better?
The consumers can wait on a conditional variable and be awaken when new work arrives.
Also the producer can wait on another one if all consumers are busy.

Reply:
There would be a race on the conditional variable check then set without additional synchronization primitives  It's not a bad idea.  I'll look into it.  Save the CPU from always running.

Reply:
@panos-lunarg , please look over the updated design and let me know what you think.

Reply:
Sorry this was a bit more complicated than I expected when I just read the code the first time.
I have pushed some tentative changes in my repo so you can take a look.
There are two conditional variables and 1 mutex.
I suggest introducing a ring buffer where the producer puts work and the consumers remove from it.
When a work slot is removed the data needs to be `std::move`d so the slot is free to be reused and the mutex to be held for as short as possible

I haven't tested my changes so most likely they won't work as expected.
Perhaps we are spending too much time in a non performance critical path and because I am to be blamed for this I have no strong opinion on this. Feel free to use whichever version you think best.

The rest of the code lgtm so you have my +1

Tentative patch: https://github.com/LunarG/gfxreconstruct/commit/b7d537070013cadddead148d9ffb4d38f411da9f

Reply:
The problem with a ring buffer with File I/O is that it itself dependent on other factors and may be completed out of order from what you define.  So just because you trigger a file write for offset 0 doesn't mean it will complete before offset 1.  So you could be waiting longer for threads than you have to with a ring buffer.  I'll look over your suggestions in the patch, though.  Thanks.

Reply:
Ok.  I built a release build and the PNG output didn't make that big of an impact.  So I think the multithreading isn't needed anyway.

Reply:
> I think the multithreading isn't needed anyway.

Sounds good.  But keep it in your back pocket - I would assume it speeds up both BMP and PNG by moving encoding, optionally compression, and I/O to another thread.

Reply:
Agreed.  Yes, it does speed it up, but it's only a big concern if you're dumping out a bunch of consecutive screenshots in a row (which most people won't do regularly).